### DISCUSSION TITLE
[Feature] Allow defining upload server endpoint

### DISCUSSION THREAD

**Discussion by ddshd** [immich-app/immich#9125] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Allow defining an alternative endpoint for uploading assets.

This is beneficial to Free Cloudflare users and others who do not want to expose their entire Immich stack to the world.

Example: `upload.domain.com` is defined somewhere (within the admin setting or on the local client [app/web/etc..]..

1. For Free Cloudflare users: `upload.domain.com` can be DNS-only proxied allowing getting around the 100MB upload limit while still protecting the front-end and other endpoints behind Cloudflare.
2. For people who don't want to expose their entire stack: They can expose only `upload.domain.com` (proxied to `/api/upload...` elsewhere) and allow clients to upload assets without being on VPNs, etc..
Beneficial for background uploads as well since some VPNs won't connect fast enough on-demand (or ever) before uploading.

In my opinion the Immich server shouldn't implement HOW the upload endpoint is proxied to the actual upload endpoint `/api/upload...`. That should be done by the user, this should only tell the clients which endpoint to call for uploads allow.

If we add this to be defined on the server-end instead of the client-end (would have to be defined on every client) maybe we can add an endpoint for `get-upload-endpoint` which can be used to get the external endpoint, if set. I'm my opinion the default endpoint `/api/upload..` should always continue to work.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by bo0tzz** [comment]
If you have a reverse proxy respond on the upload endpoint with a redirect, I would expect the app to follow that. Then you can do all these complicated shenanigans without any changes to immich.

**Reply by ddshd** [reply]
I'm not sure if the app follows redirects but if it does it would probably work for scenario (1) through some Cloudflare workers/rules since that's the edge proxy

For scenario (2) They wouldn't be able to make a call to the non-public URL before being able to obtain the redirect.

i.e.:
App in the background:
- `immich.local/api/upload...` is redirected to `external.immich/api/upload...` 
- Tries to make access to `immich.local/api/upload...` but cannot because the VPN hasn't connected yet or the device isn't on the local network and therefore the connection fails and isn't able to obtain the redirect.

Unless I misunderstood

In my imagination the app should call `get-upload-endpoint` on launch and save it for use later

**Reply by bo0tzz** [reply]
1. Why not? You could just redirect to an endpoint that is not proxied, right?
2. You could run the URL on a public endpoint but without actually serving the Immich API, just only the upload endpoint redirect.

**Reply by ddshd** [reply]
Sorry I had to setup a free CF account - I just tried with 302 redirected my `domain/api/asset/upload` to `nocf.domain/api/asset/upload` and then tried to upload an asset from the app. The upload failed and I didn't see a request on my noncf domain at all.

To me it seems like the app doesn't follow redirects.

Logs from app (this is all of it):
```
[_startUpload] Manual Upload Completed - success: 0, failed: 1
```

**Reply by ddshd** [reply]
> The http package won't support redirection for POST request.

https://stackoverflow.com/a/60957465

I guess if this is true then it makes sense

**Reply by bo0tzz** [reply]
You should use [307](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307), not 302.

**Reply by ddshd** [reply]
Just tried `307` as well didn't work on the app. Ensured I had CORS access allowed as well.

On the web (in chrome) it doesn't send the cookies as the origin would of course have to be different to get around Cloudflare. The cookie by default sets itself to browser Origin since we don't set a domain on `Set-Cookie` (good) but that also means it can't be overridden with `proxy_cookie_domain`.

I have this working on the version of the app I use (local setting). I'm sure I could probably figure out how to do it on the server but depends on if the team would allow it.
